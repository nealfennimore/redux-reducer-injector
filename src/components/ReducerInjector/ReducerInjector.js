import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { curry } from 'src/utils/fp';

/**
 * Inject reducers to run with component lifecycle
 *
 * @param {Object} options Should contain reducers as an object
 * @param {any} WrappedComponent Component to wrap
 * @returns ReducerInjector
 */
function ReducerInjectorHOC( options, WrappedComponent ) {
    const { reducers, STORE_KEY } = Object.assign( { reducers: {}, STORE_KEY: 'store' }, options );

    const componentName = WrappedComponent.displayName ||
	WrappedComponent.name ||
    'Component';

    class ReducerInjector extends Component {
        static displayName = `ReducerInjector(${ componentName })`
        static propTypes = {
            [ STORE_KEY ]: PropTypes.object // eslint-disable-line
        }
        static contextTypes = {
            [ STORE_KEY ]: PropTypes.object
        }

        reducers = reducers

        componentWillMount() {
            this.injectReducers();
        }

        get store() {
            return this.props[STORE_KEY] || this.context[STORE_KEY];
        }

        get keys() {
            return Object.keys( this.reducers );
        }

        get usableKeys() {
            const injectedKeys = Object.keys( this.store.injectedReducers );
            return this.keys.filter( ( key )=> ! injectedKeys.includes( key ) );
        }

        get usableReducers() {
            return this.usableKeys.reduce( ( result, key )=>{
                result[key] = this.reducers[key];
                return result;
            }, {} );
        }

        get hasReducers() {
            return !! this.keys.length;
        }

        injectReducers() {
            if ( this.hasReducers ) {
                this.store.injectReducers( this.usableReducers );
            }
        }

        render() {
            return (
                <WrappedComponent {...this.props} />
            );
        }
    }

    return ReducerInjector;
}

export default curry( ReducerInjectorHOC );
