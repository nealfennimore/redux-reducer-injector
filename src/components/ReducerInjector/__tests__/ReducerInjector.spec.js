import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import augmentStore from 'src/store';
import ReducerInjector from '../ReducerInjector';

class Test extends Component {
    render() {
        return 'test';
    }
}

describe( 'ReducerInjector', ()=>{
    let store, createReducer, spy, reducers, TestComponent;
    beforeEach( ()=>{
        store = {
            dispatch: jest.fn(),
            replaceReducer: jest.fn()
        };

        createReducer = jest.fn();
        augmentStore( createReducer, store );
        spy = jest.spyOn( store, 'injectReducers' );

        reducers = {
            r1: jest.fn(),
            r2: jest.fn(),
            r3: jest.fn()
        };

        TestComponent = ReducerInjector( { reducers } )( Test );
    } );

    test( 'should match snapshot', ()=>{
        const component = renderer.create(
            <TestComponent store={store} />
        );
        expect( component.toJSON() ).toMatchSnapshot();
    } );

    describe( 'lifecycle', ()=>{
        test( 'should inject reducers on mount', ()=>{
            const component = mount(
                <TestComponent store={store} />
            );

            expect( component.instance().store ).toEqual( store );
            expect( component.instance().reducers ).toEqual( reducers );
            expect( component.instance().hasReducers ).toEqual( true );
            expect( component.instance().usableReducers ).toMatchSnapshot();

            expect( spy.mock.calls ).toMatchSnapshot();
        } );
        test( 'should inject only usable reducers on mount', ()=>{
            store.injectedReducers.r1 = jest.fn();
            const component = mount(
                <TestComponent store={store} />
            );

            expect( component.instance().store ).toEqual( store );
            expect( component.instance().reducers ).toEqual( reducers );
            expect( component.instance().hasReducers ).toEqual( true );
            expect( component.instance().usableReducers ).toMatchSnapshot();

            expect( spy.mock.calls ).toMatchSnapshot();
        } );
        test( 'should not inject when no reducers', ()=>{
            TestComponent = ReducerInjector( {} )( Test );
            const component = mount(
                <TestComponent store={store} />
            );

            expect( component.instance().store ).toEqual( store );
            expect( component.instance().reducers ).toEqual( {} );
            expect( component.instance().hasReducers ).toEqual( false );
            expect( component.instance().usableReducers ).toMatchSnapshot();
            expect( spy ).not.toBeCalled();
        } );
    } );


} );
