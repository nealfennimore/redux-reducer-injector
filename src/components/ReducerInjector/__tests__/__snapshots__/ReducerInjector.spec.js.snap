// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReducerInjector lifecycle should inject only usable reducers on mount 1`] = `Object {}`;

exports[`ReducerInjector lifecycle should inject only usable reducers on mount 2`] = `
Array [
  Array [
    Object {
      "r2": [MockFunction],
      "r3": [MockFunction],
    },
  ],
]
`;

exports[`ReducerInjector lifecycle should inject reducers on mount 1`] = `Object {}`;

exports[`ReducerInjector lifecycle should inject reducers on mount 2`] = `
Array [
  Array [
    Object {
      "r1": [MockFunction],
      "r2": [MockFunction],
      "r3": [MockFunction],
    },
  ],
]
`;

exports[`ReducerInjector lifecycle should not inject when no reducers 1`] = `Object {}`;

exports[`ReducerInjector should match snapshot 1`] = `"test"`;
